yarn-project/p2p/src/msg_validators/tx_validator/factory.ts

```diff
@@ -92,3 +96,64 @@ export function createTxMessageValidators(
     },
   ];
 }
+
+export interface ValidationResult {
+  name: string;
+  isValid: TxValidationResult;
+  severity: PeerErrorSeverity;
+}
+export type ValidationOutcome = { allPassed: true } | { allPassed: false; failure: ValidationResult };
+
+export async function validateInParallel(
+  tx: Tx,
+  validators: Record<string, MessageValidator>,
+  log?: Logger,
+): Promise<ValidationOutcome> {
+  const abortController = new AbortController();
+  const txHash = await tx.getTxHash();
+
+  const asyncTasks = Object.fromEntries(
+    Object.entries(validators).map(([name, { validator, severity }]): [string, Promise<ValidationResult>] => {
+      const asyncTask = async () => {
+        try {
+          const isValid = await validator.validateTx(tx, abortController.signal);
+          return { name, isValid, severity };
+        } catch (err) {
+          if (err instanceof AbortError) {
+            log?.trace(`Validation task ${name} was aborted for tx ${txHash}`, { name, txHash });
+            return { name, isValid: { result: 'skipped' as const, reason: ['Operation aborted'] }, severity };
+          }
+
+          log?.warn(`Unexpected error running validator ${name} on ${txHash}`, { err, name, txHash });
+          return { name, isValid: { result: 'invalid' as const, reason: [String(err)] }, severity };
+        }
+      };
+
+      return [name, asyncTask()];
+    }),
+  );
+
+  const allValidations: Array<ValidationResult> = [];
+  while (Object.keys(asyncTasks).length > 0) {
+    const nextTaskFinished = await Promise.race(Object.values(asyncTasks));
+    allValidations.push(nextTaskFinished);
+    delete asyncTasks[nextTaskFinished.name];
+
+    if (nextTaskFinished.isValid.result !== 'valid') {
+      abortController.abort();
+    }
+  }
+
+  // this will select the first 'real' failure (i.e. before any timeouts)
+  const failure = allValidations.find(x => x.isValid.result !== 'valid');
```

Comment:
Shouldn't this be `x.isValid.result === 'invalid'`, so that we propagate the actual error/reason

Reply:
I could refactor it that way but the way the while-loop above is expected to work is that it sees the first failure, saves the result inside `allValidations` and cancels the work (which ultimately would add a bunch of `skipped` results).

The `find` call will find the first which is going to be the first failure.